### PR TITLE
Storage interface + minimal implementation of SQL DB storage

### PR DESCRIPTION
**Pull request by motus**
Summary of changes:
* Define base API for storing and retrieving of the experimental data;
* Implement SQL DB-based storage system. (Tested for DuckDB, SQLite3, MySQL/MariaDB, and PostgreSQL);
* Fix the config loader to simplify the instantiation of `Storage` and `Optimizer` objects from JSON config;
* Fix the `Launcher` class to support Storage instantiation;
* Change the optimization loop in `run_opt.py` to save and restore the benchmark data;
* Make global `Service` functions available in both `Storage` and `Optimizer`;
* Add unit tests for the new `Storage` functionality

### CODE REVIEW

File: conda-envs/mlos_core-3.10.yml [hunk 1]
```diff
@@ -28,5 +28,6 @@ dependencies:
     - bump2version
     - licenseheaders
     - pytest-local-badge
+    - sqlalchemy
```
Comment (motus) [inline conda-envs/mlos_core-3.10.yml:31]: conda has SQLAlchemy 1.4, and we need v 2.0, hence pip

File: mlos_bench/config/storage/sqlite3.jsonc [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// Local SQLite3 database.
+{
+    "class": "mlos_bench.storage.SqlStorage",
+
+    // Apply --config-path resolution to these config parameters:
+    "resolve_path": ["database"],
```
Comment (motus) [inline mlos_bench/config/storage/sqlite3.jsonc:6]: A new feature in config_persistence clas loader: look up for file specified in `database` parameter in `--config-path`. This works for all JSON configs including `Environment` and `Optimizer`. (The functionality is a bit intertwined with some other Storage-specific fixes to be extracted as a separate PR, though)

File: conda-envs/mlos_core-3.10.yml (hunk 1 above)
Comment (bpkroth) [reply conda-envs/mlos_core-3.10.yml:31]: Honestly not convinced conda is any better than just raw containers, but I guess we go where the devs are at.

File: conda-envs/mlos_core-3.11.yml [hunk 3]
```diff
@@ -28,5 +28,6 @@ dependencies:
     - bump2version
     - licenseheaders
     - pytest-local-badge
+    - sqlalchemy
```
Comment (bpkroth) [inline conda-envs/mlos_core-3.11.yml:31]: Suggestion to put this into the `mlos_bench/setup.py` as an `extra_requires` and it should get handled automatically in all of the conda envs then.
(`full` specifier below gets mapped to all different `extra_requires` combos for convenience).

Comment (motus) [reply conda-envs/mlos_core-3.11.yml]: done!

**Comment by motus** [comment]
This PR is superseded by #298
